### ISSUE TITLE
Help / Tips on Model Convergence

### ISSUE THREAD

**Issue by siddu9501** [Rudrabha/Wav2Lip#97]
Training on own dataset, (I have read the README)
The Wav2Lip model gets an evaluation Sync Loss of ~1.1 consistently, based on the code, the syncnet_wt does not kick in during training till the loss reaches 0.75. The generated images and target images look fairly similar even without the sync loss kicking in. Let's say we make a checkpoint here and force set syncnet_wt to 0.01 and continue training. There are artifacts generated in the image in order to reduce the sync_loss. Would this attribute an issue with SyncNet not completed training or something else to do with Wav2Lip? 

The SyncNet training looks like this.
```
Evaluating for 1400 steps
Validation Loss: 0.2739824056625366
Train Loss: 0.22405773401260376: : 1it [00:12, 12.39s/it]
Train Loss: 0.20201614499092102: : 1it [00:12, 12.21s/it]
Train Loss: 0.2009502649307251: : 1it [00:11, 11.96s/it]
```

And my Wav2Lip looks like this for the above SyncNet model. 
```
Evaluating for 500 steps
Validation L1: 0.01017100002873176, Sync loss: 1.2633369711939684
Starting Epoch: 27000
L1: 0.008067036047577858, Sync Loss: 0.0: : 1it [00:02,  2.67s/it] Starting Epoch: 27001
```

* Other changes I made. 
- Architecture changes to incorporate 30fps 
- Some batch size / multi gpu adjustments

**Comment by prajwalkr** [comment]
> * Architecture changes to incorporate 30fps

There is also a change in this function, such as choosing the right `syncnet_mel_step_size`: https://github.com/Rudrabha/Wav2Lip/blob/929e3caa247d45989e201e43fe17f19c7e99ae56/color_syncnet_train.py#L56

While training Wav2Lip as well, there might be some changes while choosing the correct audio segment. An alternative would be to just resample all your training videos to 25fps, this not needing any changes in the code. We are unsure where the mistake could lie, at this moment. Also, your own dataset needs to be sync-corrected before training any of the networks. 

 

> There are artifacts generated in the image in order to reduce the sync_loss.

In our experience, the artifacts do come initially and the L1 loss also goes higher when the sync loss is applied, but the artifacts go away as L1 loss slowly reduces.

**Comment by siddu9501** [comment]
Yes, for 30fps, instead of changing the step_size, I chose to take 6 frames to cover 0.2 seconds off a 30 fps video. syncnet_T = 6, which means the step_size would still be 16 ; And the mel chunk for any 6 consecutive frames would be `80 * (start_frame/fps) + 16` . If we do the math, all the frames are covered with appropriate mels this way. Accordingly, after stacking you will now get 18 channels instead of 15 which was one of the architecture changes in the syncnet face encoder. Would that make sense?

Handled the segmented_mels similarly so that wav2lip takes images and corresponding mel_chunks centered around the image frames.

Is there a theory behind initiating the syncwt after the loss hits the specific number "0.75" ; Or was this something that just worked for LRS2 dataset? 

The dataset is sync-corrected. 
As for the artifacts part, that statement is reassuring. I'll try to see if there is any improvement.

**Comment by prajwalkr** [comment]
> Would that make sense?

yes, look ok. 

> Or was this something that just worked for the LRS2 dataset?

Yes, it was something that worked just fine for LRS2.
